### PR TITLE
Add exception when using 'select'

### PR DESCRIPTION
**Pull request by GinnyJI**
Add `labels` field in namespace struct within Forest. When propagating
an object, we first check if a `select` annotation exists. If this
annotation exists and doesn't match the current namespace, this object
will not be propagated.

Modify the object_test.go to use `Role` instead of `Secret` in
propagating tests.

Tested: make test

### CODE REVIEW

**Comment by GinnyJI** [comment]
tracks #1123

File: incubator/hnc/internal/forest/namespace.go [hunk 1]
```diff
@@ -27,6 +27,10 @@ type Namespace struct {
 	exists               bool
 	allowCascadingDelete bool
 
+	// labels store the original namespaces's labels, and we add the propogate label to identify itself
```
Comment (yiqigao217) [inline incubator/hnc/internal/forest/namespace.go:30]: r/namespaces's/namespaces'

File: incubator/hnc/internal/forest/namespace.go [hunk 2]
```diff
@@ -27,6 +27,10 @@ type Namespace struct {
...
+	// in the exception process
+	labels map[string]string
```
Comment (yiqigao217) [inline incubator/hnc/internal/forest/namespace.go:32]: Do we store all the labels in the forest or just the propagate labels?

File: incubator/hnc/internal/reconcilers/hierarchy_config.go [hunk 3]
```diff
@@ -398,6 +404,9 @@ func (r *HierarchyConfigReconciler) syncLabel(log logr.Logger, nsInst *corev1.Na
 		anc = anc.Parent()
 		depth++
 	}
+
+	labels[propagatePrefix+nsInst.Name] = "true"
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/hierarchy_config.go:408]: I only saw `propagate.hnc.x-k8s.io/all=true` or `propagate.hnc.x-k8s.io/none=true` annotations in the [design doc](https://docs.google.com/document/d/17J8icBEDvLLoPT4kQ4ArZcCerRweDY-TpJ48DJKpHJ0/edit#). What does `propagate.hnc.x-k8s.io/current_ns_name=true` do?

File: incubator/hnc/internal/reconcilers/object_test.go [hunk 4]
```diff
@@ -53,22 +53,22 @@ var _ = Describe("Secret", func() {
...
 
 		// Create a secret that does NOT propogate to bazName
 		a := map[string]string{"propagate.hnc.x-k8s.io/select": "!propagate.hnc.x-k8s.io/" + bazName}
-		makeObjectWithAnnotation(ctx, "Secret", fooName, "fooSecret", a)
+		makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
 
-		Eventually(hasObject(ctx, "Secret", barName, "fooSecret")).Should(BeTrue())
-		Consistently(hasObject(ctx, "Secret", bazName, "fooSecret")).Should(BeFalse())
+		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue())
+		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse())
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object_test.go:71]: Just a question here, what's the difference between `Eventually` and `Consistently`?

File: incubator/hnc/internal/forest/namespace.go (hunk 1 above)
Comment (GinnyJI) [reply incubator/hnc/internal/forest/namespace.go:30]: done

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 3 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:408]: I found out that I had to add the `propagate.hnc.x-k8s.io/current_ns_name` label manually for each ns, otherwise the selector can't identify whether to propagate or not. The `true` value is just to say that this ns has the name `current_ns_name`

File: incubator/hnc/internal/reconcilers/object_test.go (hunk 4 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object_test.go:71]: `Eventually` is to say we are able to get value A at some point within a period of time. `Consistently` means we are able to get value A for a stable period of time (e.g. the `False` doesn't get to changed to `True` after like 1s or so)

Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object_test.go:71]: Thanks for the explanation!

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 3 above)
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:408]: IIUC the selector will look for `current_ns_name.tree.hnc.x-k8s.io/depth: 0` label instead of a new `propagate.hnc.x-k8s.io/current_ns_name: true` label. WDYT? @adrianludwin

Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:408]: Label design needs to be done _very_ carefully. @GinnyJI , why won't the depth labels that Yiqi mentioned work?

File: incubator/hnc/internal/reconcilers/hierarchy_config.go [hunk 5]
```diff
@@ -43,6 +43,8 @@ import (
 	"sigs.k8s.io/multi-tenancy/incubator/hnc/internal/stats"
 )
 
+const propagatePrefix = "propagate.hnc.x-k8s.io/"
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/hierarchy_config.go:46]: IMO this should go into the API package, since the prefix is part of the API.

File: incubator/hnc/internal/reconcilers/hierarchy_config.go [hunk 6]
```diff
@@ -398,6 +404,9 @@ func (r *HierarchyConfigReconciler) syncLabel(log logr.Logger, nsInst *corev1.Na
 		anc = anc.Parent()
 		depth++
 	}
+
+	labels[propagatePrefix+nsInst.Name] = "true"
+	ns.SetLabels(labels)
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/hierarchy_config.go:409]: This doesn't look correct to me. It looks like we're _only_ storing the labels that HNC itself has created, but we should be storing _all_ the labels on the namespace, including the ones set by the users.

File: incubator/hnc/internal/reconcilers/object.go [hunk 7]
```diff
@@ -53,6 +54,9 @@ const (
 	actionNop     syncAction = "no-op"
 
 	unknownSourceNamespace = "<unknown-source-namespace>"
+
+	selectorAnnot     = "propagate.hnc.x-k8s.io/select"
+	treeSelectorAnnot = "propagate.hnc.x-k8s.io/treeSelect"
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:59]: These constants should be in the API package. Also, you shouldn't be repeating yourself so much - "hnc.x-k8s.io" is already called something like `api.MetaGroup`, and you should have a constant for the `propagate.` part as well.

Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/object.go:59]: https://github.com/kubernetes-sigs/multi-tenancy/blob/master/incubator/hnc/api/v1alpha2/hierarchy_types.go#L32

File: incubator/hnc/internal/reconcilers/object.go [hunk 8]
```diff
@@ -359,18 +363,20 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
-	// source instance. Note that DeepEqual could return `true` even if the object doesn't exist if
-	// the source object is trivial (e.g. a completely empty ConfigMap).
-	if !exists ||
-		!reflect.DeepEqual(object.Canonical(inst), object.Canonical(srcInst)) ||
-		inst.GetLabels()[api.LabelInheritedFrom] != srcInst.GetNamespace() {
-		metadata.SetLabel(inst, api.LabelInheritedFrom, srcInst.GetNamespace())
-		return actionWrite, srcInst
+	selector := r.getSelector(ctx, log, srcInst)
+
+	if selector == nil || selector.Matches(labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())) {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:370]: I think an early exit might be much more appropriate here (e.g. above line 366). This is very hard to read, with two levels of nested `if` statements (and also this line is much too long, IMO - you can/should introduce some temp variables here).

Also - what happens if `srcInst` _used_ to be propagated here, but now has a new selector that prevents it from being propagated here? This will execute line 382 which says "the object already exists and doesn't need to be updated" - does that seem correct?

File: incubator/hnc/internal/reconcilers/object.go [hunk 9]
```diff
@@ -382,6 +388,24 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	return actionNop, nil
 }
 
+func (r *ObjectReconciler) getSelector(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:391]: `ctx` doesn't appear to be used in this function?

File: incubator/hnc/internal/reconcilers/object.go [hunk 10]
```diff
@@ -382,6 +388,24 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+	selectorstr, present := annot[selectorAnnot]
+	if present == false {
+		return nil
+	}
+	labelSelector, err := v1.ParseToLabelSelector(selectorstr)
+	if err != nil {
+		log.Error(err, "Could not parse selector annotation to labelSelector")
+		return nil
+	}
+	selector, err2 := v1.LabelSelectorAsSelector(labelSelector)
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:402]: No need to introduce a new name - `err` should work fine.

File: incubator/hnc/internal/reconcilers/object.go [hunk 11]
```diff
@@ -382,6 +388,24 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	return actionNop, nil
 }
 
+func (r *ObjectReconciler) getSelector(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	annot := inst.GetAnnotations()
+	selectorstr, present := annot[selectorAnnot]
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:393]: Prefer camel case (`selectorStr`). Also, if you liked, you could say:

```
selectorStr, ok := inst.GetAnnotations()[selectorAnnot]
```

The `ok` name is often used to see if an entry exists in a map.

File: incubator/hnc/internal/reconcilers/object.go [hunk 12]
```diff
@@ -382,6 +388,24 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+	if present == false {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:394]: Style: `!present`, not `present == false`

File: incubator/hnc/internal/reconcilers/object.go [hunk 13]
```diff
@@ -382,6 +388,24 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+		return nil
+	}
+	labelSelector, err := v1.ParseToLabelSelector(selectorstr)
+	if err != nil {
+		log.Error(err, "Could not parse selector annotation to labelSelector")
+		return nil
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:400]: We need a better way to surface these errors. @yiqigao217 can you please work with Ginny to determine one?

File: incubator/hnc/internal/forest/namespace.go (hunk 2 above)
Comment (GinnyJI) [reply incubator/hnc/internal/forest/namespace.go:32]: Changed to store all the labels

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 3 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:408]: As discussed offline, the test is updated and this label is removed

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 5 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:46]: Removed

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 6 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:409]: Done

File: incubator/hnc/internal/reconcilers/object.go (hunk 7 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:59]: done

File: incubator/hnc/internal/reconcilers/object.go (hunk 9 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:391]: It's also not used in `syncPropagated`, so I thought I need to keep it even though it's not used directly

File: incubator/hnc/internal/reconcilers/object.go (hunk 10 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:402]: done

File: incubator/hnc/internal/reconcilers/object.go (hunk 11 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:393]: done

File: incubator/hnc/internal/reconcilers/object.go (hunk 12 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:394]: done

File: incubator/hnc/internal/reconcilers/object.go (hunk 13 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:400]: Discussed with Yiqi offline and left a `TODO` here. Yiqi will draft a design doc and I'll implement at another PR

File: incubator/hnc/api/v1alpha2/hierarchy_types.go [hunk 14]
```diff
@@ -36,6 +36,10 @@ const (
 	FinalizerHasSubnamespace = MetaGroup + "/hasSubnamespace"
 	LabelTreeDepthSuffix     = ".tree." + MetaGroup + "/depth"
 	AnnotationManagedBy      = MetaGroup + "/managedBy"
+	AnnotationPropagate      = "propagate." + MetaGroup
```
Comment (yiqigao217) [inline incubator/hnc/api/v1alpha2/hierarchy_types.go:39]: Use `AnnotationPropagatePrefix`?

File: incubator/hnc/config/crd/bases/hnc.x-k8s.io_hierarchyconfigurations.yaml [hunk 15]
```diff
@@ -20,23 +20,18 @@ spec:
       description: Hierarchy is the Schema for the hierarchies API
       properties:
         apiVersion:
-          description: 'APIVersion defines the versioned schema of this representation
-            of an object. Servers should convert recognized schemas to the latest
-            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+          description: 'APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
```
Comment (yiqigao217) [inline incubator/hnc/config/crd/bases/hnc.x-k8s.io_hierarchyconfigurations.yaml:23]: You will need to rebase and run `make manifest` to remove the diffs on the `.yaml` files.

File: incubator/hnc/internal/forest/namespace.go [hunk 16]
```diff
@@ -27,6 +27,10 @@ type Namespace struct {
 	exists               bool
 	allowCascadingDelete bool
 
+	// labels store the original namespaces' labels, and we add the propogate label to identify itself
```
Comment (yiqigao217) [inline incubator/hnc/internal/forest/namespace.go:30]: Update the comment since we don't have propagate labels now after the first push.

File: incubator/hnc/internal/reconcilers/hierarchy_config.go [hunk 17]
```diff
@@ -377,6 +377,7 @@ func (r *HierarchyConfigReconciler) syncLabel(log logr.Logger, nsInst *corev1.Na
 	// and cycles.
 	anc := ns
 	depth := 0
+
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/hierarchy_config.go:380]: Maybe remove this line?

File: incubator/hnc/internal/reconcilers/object.go [hunk 18]
```diff
@@ -359,9 +360,20 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+
+	instLabels := labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())
+	if selector != nil && !selector.Matches(instLabels) {
+		// The object already exists and doesn't need to be updated. This will typically happen when HNC
+		// is restarted - all the propagated objects already exist on the apiserver. Record that it exists
+		// for our statistics.
+		r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
+
+		// Nothing more needs to be done.
+		return actionNop, nil
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object.go:375]: The logic seems not right to me. Why do we move this part up? I think it should be that if it doesn't match we will just early exit with `actionNop`. The rest of the logic to `actionWrite` if there's a diff and `actionNop` if there's not should remain the same.

File: incubator/hnc/internal/reconcilers/object.go [hunk 19]
```diff
@@ -370,16 +382,27 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 		!reflect.DeepEqual(object.Canonical(inst), object.Canonical(srcInst)) ||
 		inst.GetLabels()[api.LabelInheritedFrom] != srcInst.GetNamespace() {
 		metadata.SetLabel(inst, api.LabelInheritedFrom, srcInst.GetNamespace())
-		return actionWrite, srcInst
 	}
+	return actionWrite, srcInst
+}
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object.go:387]: Same as the comment above, this should still be in the if clause.

File: incubator/hnc/internal/reconcilers/object_test.go [hunk 20]
```diff
@@ -53,22 +53,22 @@ var _ = Describe("Secret", func() {
...
 	})
 
-	PIt("should propagate object only to selected namespace using select", func() {
+	FIt("should propagate object only to selected namespace using select", func() {
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object_test.go:62]: Remove the `F` in `FIt` before checking in.

File: incubator/hnc/internal/reconcilers/object.go [hunk 21]
```diff
@@ -370,16 +382,27 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
-	// The object already exists and doesn't need to be updated. This will typically happen when HNC
-	// is restarted - all the propagated objects already exist on the apiserver. Record that it exists
-	// for our statistics.
-	r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
-
-	// Nothing more needs to be done.
-	return actionNop, nil
+func (r *ObjectReconciler) getSelector(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	annot := inst.GetAnnotations()
+	selectorStr, ok := annot[api.AnnotationSelector]
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object.go:391]: Just confirming: Is it that `AnnotationTreeSelector` is not yet implemented in the PR?

Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:391]: Yes, it's not part of this pr

File: incubator/hnc/api/v1alpha2/hierarchy_types.go (hunk 14 above)
Comment (GinnyJI) [reply incubator/hnc/api/v1alpha2/hierarchy_types.go:39]: Done

File: incubator/hnc/internal/reconcilers/object.go (hunk 9 above)
Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/object.go:389]: Maybe it used to be used there but was removed. I'd take it out of any place it's not (currently) needed, it's easy to add it back.

File: incubator/hnc/internal/forest/namespace.go [hunk 22]
```diff
@@ -88,6 +92,14 @@ func (ns *Namespace) UnsetExists() bool {
...
 }
 
+func (ns *Namespace) GetLabels() map[string]string {
+	return ns.labels
+}
+
+func (ns *Namespace) SetLabels(labels map[string]string) {
+	ns.labels = labels
+}
+
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/namespace.go:102]: If we never do anything smart with these fields, maybe we should just make `ns.Labels` public?

Comment (adrianludwin) [reply incubator/hnc/internal/forest/namespace.go:102]: Or should we make a deep copy of the labels that get passed in so they can't be modified later? (`ns.labels = labels` is a shallow copy)

File: incubator/hnc/internal/reconcilers/object.go (hunk 18 above)
Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/object.go:375]: @yiqigao217 how much will this logic change with the changes you're making around object overwriting? It may not be worth fine-tuning this function as it exists today if you're about to change the way `syncObject` works.

With that said, I mostly agree with Yiqi. This should be an early-exit, but it shouldn't be a no-op - the object might exist and need to be removed. Your existing code doesn't account for this either - so you should probably add a new test which is to ensure that if you _add_ a label selector to an _existing_ object, that it gets _deleted_ from any namespace in which it no longer belongs.

File: incubator/hnc/internal/reconcilers/object.go [hunk 23]
```diff
@@ -359,9 +360,20 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	if srcInst == nil {
 		return actionRemove, nil
 	}
-
 	// If an object doesn't exist, assume it's been deleted or not yet created.
 	exists := inst.GetCreationTimestamp() != v1.Time{}
+	selector := r.getSelector(ctx, log, srcInst)
+
+	instLabels := labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:367]: This is not the `instLabels`, it's the namespace's labels. So let's call this `nsLabels`.

File: incubator/hnc/config/crd/bases/hnc.x-k8s.io_hierarchyconfigurations.yaml (hunk 15 above)
Comment (GinnyJI) [reply incubator/hnc/config/crd/bases/hnc.x-k8s.io_hierarchyconfigurations.yaml:23]: Done

File: incubator/hnc/internal/forest/namespace.go (hunk 16 above)
Comment (GinnyJI) [reply incubator/hnc/internal/forest/namespace.go:30]: Done

File: incubator/hnc/internal/reconcilers/hierarchy_config.go (hunk 17 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/hierarchy_config.go:380]: Done

**Comment by adrianludwin** [comment]
Is `pull-hnc-test` just failing because there's a focused test?

File: incubator/hnc/internal/reconcilers/object.go (hunk 23 above)
Comment (GinnyJI) [reply incubator/hnc/internal/reconcilers/object.go:367]: done

File: incubator/hnc/internal/forest/namespace.go (hunk 22 above)
Comment (GinnyJI) [reply incubator/hnc/internal/forest/namespace.go:102]: done. Made deep copies for `ns.labels`

File: incubator/hnc/internal/forest/namespace.go [hunk 24]
```diff
@@ -88,6 +91,17 @@ func (ns *Namespace) UnsetExists() bool {
 	return changed
 }
 
+func (ns *Namespace) GetLabels() map[string]string {
+	return ns.labels
+}
+
+func (ns *Namespace) SetLabels(labels map[string]string) {
+	ns.labels = make(map[string]string)
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/namespace.go:99]: Maybe just comment why we're doing a deep copy

File: incubator/hnc/internal/reconcilers/object.go [hunk 25]
```diff
@@ -359,9 +360,26 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
 	// If an object doesn't exist, assume it's been deleted or not yet created.
 	exists := inst.GetCreationTimestamp() != v1.Time{}
+	selector := r.getSelector(ctx, log, srcInst)
+
+	nsLabels := labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())
+	if selector != nil && !selector.Matches(nsLabels) {
+		if exists {
+			// The object already exists but it shouldn't be here according to the updated selector, so we remove it
+			r.deleteObject(ctx, log, inst)
+			return actionRemove, nil
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:372]: Returning `actionRemove` tells the _caller_ that it needs to delete the object, so don't call `r.deleteObject` here. We never make any commands to the apiserver while we hold the global mutex since network calls are slow (seconds) whereas we want to be able to lock/unlock the mutex hundreds of times a second.

Also, you don't have any error handling here, but `r.deleteObject()` return an error and we should never just ignore errors (without a very good reason why).

File: incubator/hnc/internal/reconcilers/object.go [hunk 26]
```diff
@@ -359,9 +360,26 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+		} else {
+			// The object already exists and doesn't need to be updated. This will typically happen when HNC
+			// is restarted - all the propagated objects already exist on the apiserver. Record that it exists
+			// for our statistics.
+			r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
+
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:378]: Why is this needed?

File: incubator/hnc/internal/reconcilers/object.go [hunk 27]
```diff
@@ -373,15 +391,29 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 		return actionWrite, srcInst
 	}
 
-	// The object already exists and doesn't need to be updated. This will typically happen when HNC
-	// is restarted - all the propagated objects already exist on the apiserver. Record that it exists
-	// for our statistics.
-	r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
-
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:380]: Why is this removed?

**Comment by GinnyJI** [comment]
Submitting new PRs instead of this one. Closing this PR.
